### ISSUE TITLE
A simple question: How could I read only partial data from "tfevents" file?

### ISSUE THREAD

**Issue by littleship123** [tensorflow/tensorboard#6388]
I used the OpenPCDet toolbox (https://github.com/open-mmlab/OpenPCDet) and obtained the evaluated results of 100 epochs using TensorboardX installed with OpenPCDet. 

What should I do if I only want to view the results of the 20 epochs from 100 epochs? There seems to be no data axis area adjustment function or local zoom button in the operation window. Specifically, I want to produce a new "tfevents" file based on the existing "tfevents" files or just read partial data from  the existing "tfevents" files.

I have other results only evaluated in 20 epochs.  If I read all these "tfevents" files together, it would be very unclear.
![image](https://github.com/tensorflow/tensorboard/assets/78254695/b23b4c02-2fc0-4dee-8695-22e53fb782f6)

A foolish approach is to re-evaluate those check-points of the 20 epochs, and the results overlap with the curves of the 100 epochs. You can see that the blue curve represents the results of the 100 epochs, while the orange curve covered by the blue curve represents the results of the 20 epochs.
![image](https://github.com/tensorflow/tensorboard/assets/78254695/423d87c8-87df-48de-ab2b-548b6f6ee602)

Obviously, if I could simply read the results of the 20 epochs I want from the "tfevents" file of 100 epochs, and if possible, produce a new "tfevents" file only containing results of the 20 epochs , it would be much simpler and more convenient.

I'm sorry that I didn't find the answer to my questions online, and I'm just a beginner without much relevant knowledge, so I have to ask for help here. I would be very grateful if someone could help me. Thanks to anyone who is willing to provide suggestions.

**Comment by yatbear** [comment]
Hi @littleship123,

As a *one-off* data processing operation, it's possible to iterate over the existing events in each `events.out.tfevents.xxx` file using TensorBoard backend processing functionalities, for example:
```
from tensorboard.backend.event_processing.event_file_loader import RawEventFileLoader
from tensorboard.compat.proto import event_pb2

loader = RawEventFileLoader("{path_to_your_logdir}/events.out.tfevents.{old_suffix}")
events = []
for raw_event in loader.Load():
   e = event_pb2.Event.FromString(raw_event)
   # Check epoch to see if you want to save this event proto.
   if e.step <= N:
     events.append(e)

# And then save them to a new file.
with open("{path_to_your_logdir}/events.out.tfevents.{NEW_suffix}", "wb") as f:
    for e in events:
       f.write(e.SerializeToString())  
```
However, the the utilities above are not intended to be public user-facing APIs and are subject to internal changes, please **avoid** directly using them as stable and long term dependencies in your checked-in code. 

Also please note that the above is just an example snippet put together in this comment directly, I'm didn't execute it to test, so it might need some tweaks. 

Thanks!

**Comment by littleship123** [comment]
Thank you very much. I have found it to be 'partially' useful. @yatbear 

It seems to have indeed read the content from the  "tfevents"  file, which also could be printed and written in a new  "tfevents"  file.
![image](https://github.com/tensorflow/tensorboard/assets/78254695/1a774269-2c6b-406d-b3e8-4eeb95b73b93)
![image](https://github.com/tensorflow/tensorboard/assets/78254695/82f12b5b-3d4e-4f93-90be-79f4ed908d1c)

However, when I try to view it using the command line, an error is reported. It looks the same as the issue #5116 . You can open a display window, but nothing will be displayed.
![image](https://github.com/tensorflow/tensorboard/assets/78254695/e0f2a82f-9b4a-4aee-b08a-34f73c97dd44)

The obtained `e` is a list, but the type of content in it seems to be special (like `<class 'google.protobuf.pyext._message.RepeatedCompositeContainer'>`). I can read `e.summary.value` like reading a list, but I cannot read `e.summary.value.tag`.

But I still managed to use some normal methods to generate new "tfevents" files. At least it seems successful now. 
![image](https://github.com/tensorflow/tensorboard/assets/78254695/7945f705-9440-4626-b9bb-bad50e9ba132)
![image](https://github.com/tensorflow/tensorboard/assets/78254695/fcc26101-6041-4ec9-adfc-9296940896b6)

I am using `SummaryWriter` in TensorBoardX for pytorch, but I think it should also be possible to be replaced with the corresponding code for TensorBoard.
I hope it can provide assistance to others with similar problem. Thank you again for the provided code. @yatbear 

'''python
from tensorboardX import SummaryWriter
from tensorboard.backend.event_processing.event_file_loader import RawEventFileLoader
from tensorboard.compat.proto import event_pb2

loader = RawEventFileLoader("{path_to_your_logdir}/events.out.tfevents.{old_suffix}.ps") # it is a file name, need to add "ps"
events = []
for raw_event in loader.Load():
    e = event_pb2.Event.FromString(raw_event)
    # print(e)
    # print(e.step) # the "step" in my result is the index of training epoch, I want to choose the last 20 epochs.
    if e.step <= 100 and e.step > 80:
            events.append(e)

tb_log = SummaryWriter(log_dir="{path_to_your_logdir}/events") # it is a path, not a file, I create a folder called events
for e in events[1:]:   # the first element in "e" is not the data, you can print "events[0:1]" to look at it
    # print(e.summary.value)
    # print(type(e.summary.value))
    for value in e.summary.value:
        # print(value.tag,value.simple_value, e.step)
        tb_log.add_scalar(str(value.tag),value.simple_value, e.step)
'''

**Comment by littleship123** [comment]
All right.
It seems that I managed to use some normal methods to generate new "tfevents" files based on `e` rather than `f.write` into the file. At least, it looks successful for now. 
![image](https://github.com/tensorflow/tensorboard/assets/78254695/7945f705-9440-4626-b9bb-bad50e9ba132)
![image](https://github.com/tensorflow/tensorboard/assets/78254695/fcc26101-6041-4ec9-adfc-9296940896b6)


I am using `SummaryWriter` in "TensorBoardX" for pytorch, but I also install "TensorFlow" and "TensorBoard" through "Conda", so I can use function from "TensorBoardX" and "TensorBoard". I think it should also be possible to be replaced with the corresponding code for "TensorBoard" . I guess their process is similar.

I hope it can provide assistance to others with similar problem. Thank you again for the provided code. @yatbear 
```python
from tensorboardX import SummaryWriter
from tensorboard.backend.event_processing.event_file_loader import RawEventFileLoader
from tensorboard.compat.proto import event_pb2

loader = RawEventFileLoader("{path_to_your_logdir}/events.out.tfevents.{old_suffix}.ps") # it is a file name, need to add "ps"
events = []
for raw_event in loader.Load():
    e = event_pb2.Event.FromString(raw_event)
    # print(e)
    # print(e.step) # the "step" in my result is the index of training epoch, I want to choose the last 20 epochs.
    if e.step <= 100 and e.step > 80:
            events.append(e)

tb_log = SummaryWriter(log_dir="{path_to_your_logdir}/events") # it is a path, not a file, I create a folder called events
for e in events[1:]:   # the first element in "e" is not the data, you can print "events[0:1]" to look at it
    # print(e.summary.value)
    # print(type(e.summary.value))
    for value in e.summary.value:
        # print(value.tag,value.simple_value, e.step)
        tb_log.add_scalar(str(value.tag),value.simple_value, e.step)
```

Additionally, a better approach may be to evaluate only 20 epochs per evaluation operation (and per Tensorboard record), rather than directly evaluating all 100 epochs like me. It could be a bit troublesome to break a "tfevents" file apart.
